### PR TITLE
Number Utilities

### PR DESCRIPTION
**Pull request by cicerorph**

### CODE REVIEW

**Comment by hammouda101010** [comment]
this should be renamed to "number formating" tbh i'm getting sick of these "utilites" names

**Comment by cicerorph** [comment]
> this should be renamed to "number formating" tbh i'm getting sick of these "utilites" names

Idc

**Comment by cicerorph** [comment]
It has some other utilities other than formatting

**Comment by cicerorph** [comment]
Fixed it lol

**Comment by JeremyGamer13** [comment]
if your gonna have error messages, i would recommend just throwing errors so users can use the native try catch blocks instead of checking if the returned value starts with error

**Comment by CodeFoxy-Github** [comment]
11:24 wtf

JeremyGamer13 ***@***.***>於 2024年10月29日 週二，下午11:23寫道：

**Comment by cicerorph** [comment]
> if your gonna have error messages, i would recommend just throwing errors so users can use the native try catch blocks instead of checking if the returned value starts with error

Updated.

File: static/extensions/MubiLop/numutils.js [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+          {
+            opcode: 'toLocaleString',
+            blockType: BlockType.REPORTER,
+            text: 'format number [NUMBER] with locale [LOCALE]',
+            arguments: {
+              NUMBER: {
+                type: ArgumentType.NUMBER,
+                defaultValue: 10000
+              },
+              LOCALE: {
```
Comment (jwklong) [inline static/extensions/MubiLop/numutils.js:116]: accept reporters?

File: static/extensions/MubiLop/numutils.js [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+          {
+            opcode: 'formatCurrency',
+            blockType: BlockType.REPORTER,
+            text: 'format [NUMBER] as currency in [CURRENCY]',
+            arguments: {
+              NUMBER: {
+                type: ArgumentType.NUMBER,
+                defaultValue: 123.45
+              },
+              CURRENCY: {
```
Comment (jwklong) [inline static/extensions/MubiLop/numutils.js:177]: accept reporters?

File: static/extensions/MubiLop/numutils.js [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+            text: 'is [NUMBER] NaN?',
+            arguments: {
+              NUMBER: {
+                type: ArgumentType.NUMBER,
+                defaultValue: 'NaN'
+              }
+            }
+          },
+          {
+            opcode: 'parseNumber',
```
Comment (jwklong) [inline static/extensions/MubiLop/numutils.js:248]: numbers already get parsed by scratch automatically, giving this block no purpose

File: static/extensions/MubiLop/numutils.js [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+        ]
+      };
+    }
+
+    toLocaleString(args) {
+      try {
+        const locale = args.LOCALE === 'system' ? this.systemLocale : args.LOCALE;
+        return Number(args.NUMBER).toLocaleString(locale);
+      } catch (e) {
+        throw new Error('Error: Invalid number or locale');
```
Comment (jwklong) [inline static/extensions/MubiLop/numutils.js:286]: if the locale is invalid just make it use system locale

File: static/extensions/MubiLop/numutils.js [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+      }
+    }
+
+    toFixed(args) {
+      try {
+        return Number(args.NUMBER).toFixed(args.DECIMALS);
+      } catch (e) {
+        throw new Error('Error: Invalid input');
```
Comment (jwklong) [inline static/extensions/MubiLop/numutils.js:294]: args.decimals needs to be converted to number, making invalid input pointless

File: static/extensions/MubiLop/numutils.js [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+      }
+    }
+
+    toPrecision(args) {
+      try {
+        return Number(args.NUMBER).toPrecision(args.PRECISION);
```
Comment (jwklong) [inline static/extensions/MubiLop/numutils.js:300]: args.precision needs to be converted to number, making invalid input pointless

File: static/extensions/MubiLop/numutils.js [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+      } catch (e) {
+        throw new Error('Error: Invalid input');
+      }
+    }
+
+    toExponential(args) {
+      try {
+        return Number(args.NUMBER).toExponential(args.DECIMALS);
```
Comment (jwklong) [inline static/extensions/MubiLop/numutils.js:308]: args.decimals needs to be converted to number, making invalid input pointless

File: static/extensions/MubiLop/numutils.js [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+    }
+
+    formatCurrency(args) {
+      try {
+        return Number(args.NUMBER).toLocaleString('en-US', {
+          style: 'currency',
+          currency: args.CURRENCY
+        });
+      } catch (e) {
+        throw new Error('Error: Invalid input');
```
Comment (jwklong) [inline static/extensions/MubiLop/numutils.js:321]: just make it use like,,, system currency or whatever

File: static/extensions/MubiLop/numutils.js [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+      }
+    }
+
+    formatPercentage(args) {
+      try {
+        const number = Number(args.NUMBER);
+        return `${(number * 100).toFixed(args.DECIMALS)}%`;
+      } catch (e) {
+        throw new Error('Error: Invalid input');
```
Comment (jwklong) [inline static/extensions/MubiLop/numutils.js:330]: args.decimals needs to be converted to number, making invalid input pointless

File: static/extensions/MubiLop/numutils.js [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+      }
+    }
+
+    roundNumber(args) {
+      try {
+        const multiple = Number(args.MULTIPLE);
+        return Math.round(args.NUMBER / multiple) * multiple;
+      } catch (e) {
+        throw new Error('Error: Invalid input');
```
Comment (jwklong) [inline static/extensions/MubiLop/numutils.js:339]: args.number needs to be converted to number, making invalid input pointless

File: static/extensions/MubiLop/numutils.js [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+    }
+
+    isNaN(args) {
+      return Number.isNaN(Number(args.NUMBER));
+    }
+
+    parseNumber(args) {
+      try {
+        const parsed = parseFloat(args.TEXT);
+        return Number.isNaN(parsed) ? 'Error: Invalid number' : parsed;
```
Comment (jwklong) [inline static/extensions/MubiLop/numutils.js:358]: if its NaN just return NaN no big deal you have a block to check if it is just let people use that if they dont want NaN

File: static/extensions/MubiLop/numutils.js [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+      } catch (e) {
+        throw new Error('Error: Invalid input');
```
Comment (jwklong) [inline static/extensions/MubiLop/numutils.js:360]: not possible i believe

File: static/extensions/MubiLop/numutils.js [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+    }
+
+    clamp(args) {
+      try {
+        const num = Number(args.NUMBER);
+        const min = Number(args.MIN);
+        const max = Number(args.MAX);
+        return Math.min(Math.max(num, min), max);
+      } catch (e) {
+        throw new Error('Error: Invalid input');
```
Comment (jwklong) [inline static/extensions/MubiLop/numutils.js:371]: not possible

File: static/extensions/MubiLop/numutils.js [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+      }
+    }
+
+    clamp(args) {
+      try {
+        const num = Number(args.NUMBER);
+        const min = Number(args.MIN);
+        const max = Number(args.MAX);
```
Comment (jwklong) [inline static/extensions/MubiLop/numutils.js:368]: if min > max swap the two around

**Review by jwklong — CHANGES_REQUESTED** [review]
a feww changes that would make this better

**Comment by RedMan13** [comment]
also just as a global note: your supposed to use Scratch.Cast.ToNumber() not Number()
also none of the inputs tell pm to not type cast them so all of the inputs are getting casted twice for no reason

**Comment by cicerorph** [comment]
Just made cast work

**Comment by RedMan13** [comment]
thats only part of what i meant.........
the casts are a waste of time all together because pm itself is performing the typecasts automatically for you.............

**Comment by hammouda101010** [comment]
> thats only part of what i meant......... the casts are a waste of time all together because pm itself is performing the typecasts automatically for you.............

that's not a problem. **Turbowarp** doesn't do this by default so you'll have to do this by yourself in your extensions to make it compatible or atleast make it work.

**Comment by RedMan13** [comment]
> > thats only part of what i meant......... the casts are a waste of time all together because pm itself is performing the typecasts automatically for you.............
> 
> that's not a problem. **Turbowarp** doesn't do this by default so you'll have to do this by yourself in your extensions to make it compatible or atleast make it work.

aut this is the **PenguinMod** extension gallery

**Comment by hammouda101010** [comment]
> > > thats only part of what i meant......... the casts are a waste of time all together because pm itself is performing the typecasts automatically for you.............
> > 
> > 
> > that's not a problem. **Turbowarp** doesn't do this by default so you'll have to do this by yourself in your extensions to make it compatible or atleast make it work.
> 
> aut this is the **PenguinMod** extension gallery

however you got a point. this is **NOT** the **Turbowarp** extension gallery. I was just talking abt compability.

File: static/extensions/MubiLop/numutils.js [hunk 15]
```diff
@@ -281,97 +259,93 @@
...
+        
+        // swap stuff
+        if (min > max) {
+          [min, max] = [max, min];
+        }
+        
         return Math.min(Math.max(num, min), max);
       } catch (e) {
-        throw new Error('Error: Invalid input');
+        return 'Error: Invalid input';
```
Comment (jwklong) [inline static/extensions/MubiLop/numutils.js:345]: error not possible

File: static/extensions/MubiLop/numutils.js [hunk 16]
```diff
@@ -281,97 +259,93 @@
...
+      return Number.isInteger(Cast.toNumber(args.NUMBER));
     }
 
     isFinite(args) {
-      return Number.isFinite(Number(args.NUMBER));
-    }
-
-    isNaN(args) {
-      return Number.isNaN(Number(args.NUMBER));
-    }
```
[the remark this answers is not in the document]
Comment (cicerorph) [reply static/extensions/MubiLop/numutils.js]: Bcs it's no use anyway

[the remark this answers is not in the document]
Comment (ddededodediamante) [reply static/extensions/MubiLop/numutils.js]: You can use it to see if a variable is a number or not 
so useful

File: src/lib/Documentation/NumberUtilities.md [hunk 17]
```diff
@@ -149,11 +149,7 @@ Examples:
 - `-10` between `0` and `100` → `0`
 
 ## Error Handling
```
Comment (jwklong) [inline src/lib/Documentation/NumberUtilities.md:151]: jus a suggestion but this doesn't really seemmmm useful anymore

**Review by jwklong — APPROVED** [review]
yeahh its good i guess

**Comment by jwklong** [comment]
> > thats only part of what i meant......... the casts are a waste of time all together because pm itself is performing the typecasts automatically for you.............
> 
> that's not a problem. **Turbowarp** doesn't do this by default so you'll have to do this by yourself in your extensions to make it compatible or atleast make it work.

boo hoo

**Pull request merged**